sklearn/metrics/_classification.py

```diff
@@ -140,6 +145,33 @@ def _check_targets(y_true, y_pred):
     return y_type, y_true, y_pred
 
 
+def _weighted_sum_1d(sample_score, sample_weight, normalize=False, xp=None):
+    """Specialized _array_api._average for the 1D output case.
+
+    Implements numpy-specific variants when the backing data is managed by
+    NumPy and delegates to the generic Array API `_average` function
+    otherwise.
+
+    This function converts the result to a Python `float` to make the result
+    array namespace and device agnostic.
+    """
+    if xp is None:
+        xp, _ = get_namespace(sample_score, sample_weight)
+
+    if not _is_numpy_namespace(xp):
+        return float(_average(sample_score, weights=sample_weight, normalize=normalize))
+
+    # faster, simpler track for numpy namespace, compared to _average.
```

Comment:
I don't like that we are changing average here. But even if we want to do this, this logic should be in `_average`, not here.

Reply:
`np.average` already has a branch for `if weights is None`, not sure how much we're gaining here for the branch.

Overall, this seems like the kind of optimization that should go to `array-api-strict`

cc @rgommers

Reply:
`array-api-strict` is for testing compliance only, so any kind of performance optimization is not relevant / out of scope there.

If you meant `array-api-compat` instead: also not relevant, since `average` isn't even in the array API standard (and `mean` and other reductions don't have weights).

I'm missing some context for this PR, but fast paths for things like weighted sums seem to belong in the library that wants the fast path. In case we accumulate similar things in for example scikit-learn and SciPy, maybe we could factor them out. But this is pretty specific code.

Reply:
From a code organization point of view, we could move this branch of `_weighted_sum_1d` into a subbranch of `_average` that would be special-cased for `1d` data.

However, `np.average` does not provide a path for "unormalized" weighted sum which we need for some classification metric. If we had a `normalize=False` option to our `_average` we would not longer have the simple story that `_average` should only be an Array API compatible re-implementation of a subset of `np.average`.

Reply:
For the record, this `normalize=False` 1d weighted sum fast path was introduced 10 years ago in https://github.com/scikit-learn/scikit-learn/pull/3531/files#diff-03d2599d34b3975da0e0aed9b1f0b4795c46eb0243d6bc4a5cae904b5f1c68a2R95 and is used only in classification metrics. For the record, it seems to still be a valid optim:

```python
>>> import numpy as np
>>> data = np.random.randn(int(1e8))
>>> sample_weight = np.random.uniform(size=int(1e8))
>>> %timeit np.average(data, weights=sample_weight) * sample_weight.shape[0]
167 ms ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.dot(data, sample_weight)
29.9 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

In the context of the Array API PRs, we tried not to change the behavior of the existing scikit-learn code base.

Reply:
Ok I merged the `np.dot` optim into `_average`.